Integrate Vehicle Model Generation and kuksa-java-sdk

wba2hi opened the pull request:
Contains two bugfixes which were required for Java:
vss-processor-plugin -> VssProcessorPlugin.kt => use JavaCompile instead of preCompile (the later is an android specific gradle task)
vss-processor -> VssNodeSpecModel.kt => Change equals and hashCode method to use the path instead of the uuid. The uuid is not part of all (especially BYD-specific) VSS Specs. However the (absolute) path itself should be a sufficient id, because it will not contain two nodes which are named "Vehicle.Speed" or comparably.

wba2hi committed:
feature: Add vss-processor-plugin and vss-processor Modules

wba2hi committed:
docs: Add Samples

wba2hi committed:
chore: Change Package from kuksa to velocitas

wba2hi committed:
chore: Add Repository for Sonatype

wba2hi committed:
docs: Add Documentation about VSS Model Generation

wba2hi committed:
feature: Integrate kuksa-java-sdk

wba2hi commented on sdk/src/main/java/org/eclipse/velocitas/sdk/databroker/v2/DataBrokerConnectorV2.kt:
```diff
@@ -0,0 +1,87 @@
+ * The DataBrokerConnector is used to establish a successful connection to the DataBroker. The communication takes
+ * place inside the [managedChannel]. Use the [defaultDispatcher] for the coroutine scope.
+ */
+class DataBrokerConnectorV2 @JvmOverloads constructor(
```
Really struggling with this implementation here, because it basically belongs directly into Kuksa instead of into the vehicle-app-java-sdk. But when I move it there, I have to align a lot of stuff (e.g. v1 / v2 specific package names, interfaces, add Vehicle Model Support etc) which feels too much for this ticket itself. 

However the outcome of this monstrosity is now that we have two different implementations for kuksa, kuksa.v1 from the kuksa-java-sdk and kuksa.v2 inside the vehicle-app-java-sdk which are both using different stubs inside (e.g. kuksa.v1 is implemented using blocking and async stubs while kuksa.v2 is implemented using the coroutine-stub). 
My personal preference would have been to change everything to Futures to have an improved Java Support but I wanted to keep them aligned. Changing kuksa.v1 to Futures also would generate a lot of follow-up work migrating the stuff inside the testapp etc

Therefore: 

Decision for now: Implement DataBrokerConnectorV2 loosely inside vehicle-app-java-sdk and later on move it inside kuksa-java-sdk and make it clean (e.g. one DataBrokerConnection which differentiates between kuksa.v1 and kuksa.v2 protocols etc)

wba2hi replied:
Decided to move the logic directly inside kuksa-java-sdk. It won't cover all things and will mostly be just a side-by-side implementation instead of having shared interfaces and common classes, but it's still better over there than inside this repository.

wba2hi commented on vss-processor-plugin/src/main/java/org/eclipse/velocitas/vssprocessor/plugin/VssProcessorPlugin.kt:
```diff
@@ -0,0 +1,155 @@
+                    outputDir.set(vssFile)
+                }
+
+            tasks.withType<JavaCompile> {
+                dependsOn(provideVssFilesTask.get())
+            }
```
was previously "tasks.getByName("preBuild")" however preBuild is an android specific callback

wba2hi commented on vss-processor/src/main/kotlin/org/eclipse/velocitas/vssprocessor/spec/VssNodeSpecModel.kt:
```diff
@@ -0,0 +1,392 @@
+            .build()
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (other !is VssNodeSpecModel) return false
+
+        return vssPath == other.vssPath
+    }
+
+    override fun hashCode(): Int {
+        return vssPath.hashCode()
+    }
```
was previously using uuid. However uuid is not mandatory to be set (especially when dealing with the byd sample vss specs)

wba2hi commented:
Moved all kuksa.val.v2 related stuff to kuksa-java-sdk, so this PR needs to be merged first: https://github.com/eclipse-kuksa/kuksa-java-sdk/pull/8

wba2hi committed:
chore: Remove kuksa.val.v2 Protos and Protobuf related Code

wba2hi committed:
chore: Adapt Documentation

wba2hi committed:
chore: Use Release Version of kuksa-java-sdk:0.3.1

wba2hi committed:
chore: Rename Offline Artifact

wba2hi committed:
chore: Add Dash Check

wba2hi committed:
chore: Remove Workflow breaking Copyright

Chrylo commented:
Could we add a run config(s) for the tests? :)

Chrylo commented on README.md:
````diff
@@ -7,8 +7,222 @@ This project is in incubation status. Not all required functionality might be mi
+}
+```
+
+Use the [`VssModelGenerator`](https://github.com/eclipse-kuksa/kuksa-android-sdk/blob/main/vss-core/src/main/java/org/eclipse/kuksa/vsscore/annotation/VssModelGenerator.kt) annotation.
````
Outdated?

wba2hi replied:
Well,... that's a good question indeed :D 

I didn't touch the annotation because it was part of the vss-core, to be consistent the better approach would probably be to remove the annotation from vss-core and move it into the vehicle-app-java-sdk. However if we do so, we also have to add the vehicle-app-java-sdk as a dependency to the vss-processor which isn't good. 

The cleanest solution would probably be to create a velocitas-core where the annotation is put into, but for one class...? What do you think?

Chrylo replied:
I would prefer the addition module with currently only one class.

wba2hi replied:
I removed the vss-processor and merged it's logic with the vss-processor-plugin therefore there is no need for the annotation anymore, please test this commit extra thoroughly. 

I already tested different use-cases like:
- use `vss_rel_4.0.yaml` and make `./gradlew clean build`
- add new branch + signal to `vss_rel_4.0.yaml` and execute `./gradlew build` => expectation: new file is generated for added branch, the signal is correctly added to the branch
- remove previously added branch and execute `./gradlew build` => expectation: previously existing VssModel is removed
- remove `vss_rel_4.0.yaml" and execute `./gradlew build` => build fails with meaningful error

and stuff like this, but please test twice and extra thoroughly :D

Chrylo commented on vss-processor-plugin/build.gradle.kts:
```diff
@@ -0,0 +1,172 @@
+// caching so the cache does not know about this. The issue will be ignored as a warning for now.
+//
+// Similar issue: https://github.com/gradle/gradle/issues/27940
+val pluginDescription = "Vehicle Signal Specification (VSS) Plugin of the KUKSA SDK. This is used in combination " +
```
Needs to be updated

Chrylo commented on test-core/src/main/java/org/eclipse/velocitas/test/kotest/KotestProjectConfig.kt:
Is the test-core package necessary?

wba2hi replied:
It looks like effectively the test-core is only used by the vss-processor atm (more specifically the "TestResourceFile". 

I guess we could go with both approaches here. 
a) use core-test in all modules (which saves us from having multiple "Test-Tag"-related classes as well as having multiple KotestProjectConfigs or
b) remove the core-test module and accept having some resources redundant but therefore reducing the overall complexity of the project by removing another module

Chrylo replied:
I am fine with a

wba2hi replied:
test-core has been removed (the processor module has been removed) due to the fact that it can not be used from inside vss-processor-plugin (due to it basically being a standalone rootProject) and therefore the sdk being the only consumer of the module.

Chrylo commented:
https://www.conventionalcommits.org/en/v1.0.0/ is used but the tooling is not there yet. It is fine if this is done at a later point. However two commits here are not consistent in the history.

wba2hi committed:
feature: Remove vss-processor and move Code Generation to vss-processor-plugin

wba2hi committed:
chore: Remove test-core and align Tests

wba2hi committed:
chore: Add Run Configurations for Tests and Sample App

wba2hi committed:
chore: Align OSSRH Organisational Secret Variables

see: https://github.com/eclipse-velocitas/.eclipsefdn/blob/main/otterdog/eclipse-velocitas.jsonnet#L21

Chrylo commented on vss-processor-plugin/build.gradle.kts:
```diff
@@ -27,16 +27,15 @@ plugins {
 
 // TODO: This linking to the version file currently throws a warning:
-// Caught exception: Already watching path: kuksa-android-sdk/vss-processor-plugin/..
+// Caught exception: Already watching path: vehicle-app-java-sdk/vss-processor-plugin/..
```
There is even more android stuff. See settings.gradle.kts. Do we still need that?

wba2hi replied:
Unfortunately yes, settings.gradle.kts is still needed because basically the vss-processor-plugin is still a rootProject and not a subproject (which is why we have this gradle task linking and stuff in place).

Chrylo commented on vss-processor-plugin/build.gradle.kts:
```diff
@@ -52,7 +51,7 @@ gradlePlugin {
             implementationClass = "org.eclipse.velocitas.vssprocessor.plugin.VssProcessorPlugin"
             displayName = "VSS Processor Plugin"
-            tags.set(listOf("Velocitas", "Vehicle Signal Specification", "VSS", "Android", "Kotlin"))
+            tags.set(listOf("Velocitas", "Vehicle Signal Specification", "VSS", "Java", "Android", "Kotlin"))
```
Should we remove the Android tag?

Chrylo commented on vss-processor-plugin/src/main/java/org/eclipse/velocitas/vssprocessor/plugin/VssProcessorPlugin.kt:
```diff
@@ -123,6 +108,8 @@ private abstract class ProvideVssFilesTask : DefaultTask() {
     @get:OutputDirectory
     abstract val outputDir: DirectoryProperty
 
+    private val vssModelGenerator = VssModelGenerator(project, logger)
```
Can we do this lazy? It is only needed in some cases.

Chrylo commented on vss-processor-plugin/src/main/java/org/eclipse/velocitas/vssprocessor/plugin/VssProcessorPlugin.kt:
```diff
@@ -141,9 +128,24 @@ private abstract class ProvideVssFilesTask : DefaultTask() {
+                        .filter { validVssExtension.contains(it.extension) }
+                        .toSet()
+                    if (vssFiles.isEmpty()) {
+                        logger.error("No VSS files were found! Is the plugin correctly configured?")
```
Should we return here?

wba2hi replied:
Yes, good catch

Chrylo commented on README.md:
The README still mentions symbol processing.

Chrylo commented on vss-processor-plugin/src/main/java/org/eclipse/velocitas/vssprocessor/plugin/VssProcessorPlugin.kt:
```diff
@@ -57,53 +56,39 @@ internal constructor(objectFactory: ObjectFactory) {
+            val mainSourceSet = sourceSets.getByName("main")
+            mainSourceSet.java.srcDirs(vssModelGenerator.sourceSetBasePath)
+
+            val generateVssModelsTask = project.tasks.register<GenerateVssModelsTask>("generateVssModels") {
```
Can we extract the generateVssModels to a const?

wba2hi committed:
chore: Remove unused Kotest Tags

wba2hi committed:
chore: Adapt Description in Readme.md

wba2hi committed:
chore: Remove redundant Plugin Tag

wba2hi committed:
chore: Minor Code Changes

wba2hi committed:
chore: Add GeneratedCode Path to main SrcSets for Android Projects

wba2hi committed:
chore: Allow Android-specific "check" Task to be optional

wba2hi committed:
chore: Add different Sample Projects

BLub

wba2hi committed:
chore: Exclude redundant Annotations Api from SDK

This fixes a duplicate class error which occurs due to a clash with
org.apache.tomcat:annotations-api and javax:annotations-api

wba2hi committed:
chore: Adapt Tests

wba2hi committed:
chore: Disable Functional Tests

wba2hi committed:
chore: Re-Generate .lockfiles

wba2hi commented:
Functional Tests are still failing, now however with a different error message, which doesn't make sense: 

```
* What went wrong:
A problem occurred configuring project ':lib'.
> Extension of type '**LibraryExtension**' does not exist. Currently registered extension types: [ExtraPropertiesExtension, LibrariesForLibs, VersionCatalogsExtension, BasePluginExtension, DefaultArtifactPublicationSet, SourceSetContainer, ReportingExtension, JavaToolchainService, JavaPluginExtension, **LibraryExtension**, LibraryAndroidComponentsExtension, NamedDomainObjectContainer<BaseVariantOutput>, VssProcessorPluginExtension]
```

I tried the following things:
- add more kotlin / android related libraries as dependencies
- add the libs from above to the buildSrc directory
- executing the functionalTest project manually which interestingly worked but lead to a lot of uncleaned gradle .lock files which made issues on consecutive runs and had to be removed manually
- using different packages for AppExtension and LibraryExtension

Only solution I saw was to add more sample projects for the different projects (e.g. android-app, android-lib, java-app, java-lib) and add the plugin in there.

wba2hi committed:
chore: Delete and Re-Generate gradle .lockfiles

Chrylo reviewed (approved):
LGTM

MP91 reviewed (approved):
LGTM (looks good to Mark)
